KOGITO-9463: Create a new guide describing the Job Services Quarkus Extensions (#408)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/job-services/quarkus-extensions.adoc

[#job-service-quarkus-extensions]
= Job Service Quarkus Extensions
:compat-mode!:
// Metadata:
:description: Job Service Quarkus extensions in {product_name}
:keywords: sonataflow, workflow, serverless, job service, quarkus extensions
// links
:jobs_service_image_url: https://quay.io/repository/kiegroup/kogito-jobs-service-allinone
:jobs_service_image_usage_url: https://github.com/kiegroup/kogito-images#jobs-services-all-in-one
:knative_eventing_url: https://knative.dev/docs/eventing/
:knative_eventing_trigger_url: https://knative.dev/docs/eventing/triggers/
:knative_eventing_sink_binding_url: https://knative.dev/docs/eventing/sinks/#sink-parameter-example
:quarkus_infinispan_client_reference_url: https://quarkus.io/guides/infinispan-client-reference
:quarkus_kafka_url: https://quarkus.io/guides/kafka

> :quarkus_config_reference_url: https://quarkus.io/guides/config-reference

Reviewer: I think those could be defined into the `antora.yaml` file instead of repeating in each adoc file

Author: hmmm, these are not used on this document.

For example, you can select if your workflows must interact with the Job Service by sending cloud events over the <<kogito-addons-quarkus-jobs-knative-eventing, knative eventing>> system or the <<kogito-addons-quarkus-jobs-messaging, kafka messaging>> system, or simply by executing direct <<kogito-addons-quarkus-jobs-management, REST>> calls.

Finally, for the interaction work, you must configure your Quarkus Workflow Project with the extension of your choice.

image::job-services/Quarkus-Workflow-Project-And-Extension.png[]

We recommend that you follow this procedure:

1. Identify the communication alternative that best fits your scenario.
2. Be sure that the Job Service is properly configured to support that alternative. This is very important if you want to use xref:job-services/core-concepts.adoc#knative-eventing[knative events] or xref:job-services/core-concepts.adoc#kafka-messaging[kafka messages] to communicate with it.
3. Configure your Quarkus Workflow Project with the corresponding extension.

[NOTE]
====

> If your workflows are not using timer-based actions, like timeouts, there is no need to add such extension.

Reviewer 2: ```suggestion
If your workflows are not using timer-based actions, like timeouts, there is no need to add such an extension.
```

[source,properties]
----
mp.messaging.outgoing.kogito-job-service-job-request-events.connector=quarkus-http
mp.messaging.outgoing.kogito-job-service-job-request-events.url=${K_SINK:http://localhost:8280/v2/jobs/events}
mp.messaging.outgoing.kogito-job-service-job-request-events.method=POST
----

[NOTE]
====
The `K_SINK` environment variable is automatically generated by the combination of the knative ecosystem and the SinkBinding definition that will be automatically generated in the `kogito.yml` file.

If this variable is not present, the default value `http://localhost:8280/v2/jobs/events` is used instead, this can be useful in development environments if you are executing the Job Service as a standalone service.
====

[start=2]
. Build your project and locate the automatically generated `kogito.yml` and `knative.yml` files in the `/target/kubernetes` directory of your project, xref:eventing/consume-produce-events-with-knative-eventing.adoc#proc-generating-kn-objects-build-time[see].

[source,shell]
----
mvn clean install
----

[start=3]
. Use the generated files to deploy your workflow application in the Kubernetes cluster using the following commands:

[source, bash]
----

> kubectl apply -f /target/kogito.yml
> 
> kubectl apply -f /target/knative.yml

Reviewer: ```suggestion
kubectl apply -f target/kogito.yml

kubectl apply -f target/knative.yml
```

----

You can see a full example of this interaction mode configuration in the xref:use-cases/timeout-showcase-example.adoc#execute-quarkus-project-standalone-services[Quarkus Workflow Project with standalone services] example project.

[#kogito-addons-quarkus-jobs-messaging]
== Kafka messaging interaction

To interact with the Job Service by sending cloud events over the kafka messaging system you must follow these steps:

. Be sure that you have read the xref:eventing/consume-producing-events-with-kafka.adoc[Consuming and producing events with Kafka] guide, and you have configured the project accordingly.

> . Add the `quarkus-smallrye-reactive-messaging-kafka` and `kogito-addons-quarkus-jobs-messaging` extensions to your Quarkus Workflow Project using any of the following alternatives.

Reviewer 2: ```suggestion
. Add the `quarkus-smallrye-reactive-messaging-kafka` and `kogito-addons-quarkus-jobs-messaging` extensions to your Quarkus Workflow project using any of the following alternatives.
```

Author: @kaldesai  I think "Quarkus Workflow Project" is being used all the time, and I think is good, represents a particular subject in this context ¿? Not a problem to change, but why is "Project" not good?

mp.messaging.outgoing.kogito-job-service-job-request-events.value.serializer=org.apache.kafka.common.serialization.StringSerializer
----

[start=4]
. Build and deploy your workflow application using any of the available procedures.

[#kogito-addons-quarkus-jobs-management]
== REST call interaction

To interact with the Job Service by executing direct REST calls you must follow these steps:

. Add the `kogito-addons-quarkus-jobs-management` extension to your Quarkus Workflow Project using any of the following alternatives.

[tabs]
====
Manually::
+
[source,xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-jobs-management</artifactId>
</dependency>
----
Apache Maven::
+
[source,shell]
----
mvn quarkus:add-extension -Dextensions="kogito-addons-quarkus-jobs-management"
----
Quarkus CLI::
+
[source,shell]
----
quarkus extension add kogito-addons-quarkus-jobs-management
----
====

[start=3]
. Add the following configuration to the `application.properties` file of your project.

[source,properties]
----
kogito.jobs-service.url=http://localhost:8280
----

[NOTE]
====

> When you deploy you project in a Kubernetes cluster, the `kogito.jobs-service-url` must be configured with the cloud url of the Job Service.

Reviewer 2: ```suggestion
When you deploy your project in a Kubernetes cluster, you must configure the `kogito.jobs-service-url` with the cloud URL of the Job Service.
```

====

[start=4]
. Build and deploy your workflow application using any of the available procedures.

== Job Service Embedded

To facilitate the development and testing stage of your workflows, this extension provides an embedded Job Service instance that executes in the same runtime as your workflows, and thus, requires no additional configurations. The only consideration is that it must not be used for production installations.

To use this extension you must:

. Add the `kogito-addons-quarkus-jobs-service-embedded` extension to your Quarkus Workflow Project using any of the following alternatives.

[tabs]
====
Manually::
+
[source,xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-jobs-service-embedded</artifactId>
</dependency>
----
Apache Maven::
+
[source,shell]
----
mvn quarkus:add-extension -Dextensions="kogito-addons-quarkus-jobs-management"
----
Quarkus CLI::
+
[source,shell]
----
quarkus extension add kogito-addons-quarkus-jobs-management
----
====

[start=3]
. Build and deploy your workflow application using any of the available procedures.

> You can see a full example of Job Service embedded usage in the xref:use-cases/timeout-showcase-example.adoc#execute-quarkus-project-embedded-services[Quarkus Workflow Project with embedded services] example project.

Reviewer: Please add the `Found an issue?` section at the end
